<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ hwglabr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>
        Reference
        <small>version&nbsp;0.3</small>
      </h1>
    </div>

    <div class="contents">
      <table class="ref-index">

      <colgroup>
        <col class="alias" />
        <col class="title" />
      </colgroup>

      <tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
            <p class="section-desc"></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="chr_coverage_plot.html">chr_coverage_plot</a></code> </p>
          </td>
          <td><p>Average chromosome signal (coverage) dot plot</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="chr_coverage.html">chr_coverage</a></code> </p>
          </td>
          <td><p>Average signal per chromosome</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="facs_density_plot.html">facs_density_plot</a></code> </p>
          </td>
          <td><p>FACS density plot</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="FE2.html">FE2</a></code> </p>
          </td>
          <td><p>Microarray probe log2 ratios</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="genome_average.html">genome_average</a></code> </p>
          </td>
          <td><p>Calculate genome-wide average signal</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gff_get_attribute.html">gff_get_attribute</a></code> </p>
          </td>
          <td><p>Parse gff field</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gff_read.html">gff_read</a></code> </p>
          </td>
          <td><p>Read gff file</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p></p>
          </td>
          <td><p>hwglabr: A compilation of functions used frequently by Hochwagen lab members.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="opening_act_batch_run.html">opening_act_batch_run</a></code> </p>
          </td>
          <td><p>Batch run of <code><a href='reference/opening_act.html'>opening_act</a></code> function</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="opening_act.html">opening_act</a></code> </p>
          </td>
          <td><p>Standard analysis of ChIP-seq experiments</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="order_chromosomes.html">order_chromosomes</a></code> </p>
          </td>
          <td><p>Order chromosome column by chromosome length</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_gene_arrow.html">plot_gene_arrow</a></code> </p>
          </td>
          <td><p>Plot genes on an R wiggle plot</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="readall_tab.html">readall_tab</a></code> </p>
          </td>
          <td><p>Load wiggle data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="remove_centromeres.html">remove_centromeres</a></code> </p>
          </td>
          <td><p>Remove centromeric regions from wiggle data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="rna_seq_analysis.html">rna_seq_analysis</a></code> </p>
          </td>
          <td><p>Wrapper function to run edgeR analysis of summarized reads from RNA-seq experiment</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_conv_average_plot.html">signal_at_conv_average_plot</a></code> </p>
          </td>
          <td><p>Line plot of average signal between convergent genes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_conv.html">signal_at_conv</a></code> </p>
          </td>
          <td><p>Signal between all convergent genes genome-wide</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_intergen_average_plot.html">signal_at_intergen_average_plot</a></code> </p>
          </td>
          <td><p>Line plot of average signal on intergenic regions</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_intergen.html">signal_at_intergen</a></code> </p>
          </td>
          <td><p>Signal at intergenic regions between convergent, divergent or tandem genes genome-wide</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_orf_average_plot.html">signal_at_orf_average_plot</a></code> </p>
          </td>
          <td><p>Line plot of average signal on ORFs</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_orf_se.html">signal_at_orf_se</a></code> </p>
          </td>
          <td><p>Signal around ORF start or end positions</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_orf.html">signal_at_orf</a></code> </p>
          </td>
          <td><p>Signal at all ORFs genome-wide (meta ORF)</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_rDNA.html">signal_at_rDNA</a></code> </p>
          </td>
          <td><p>Signal flanking rDNA</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_at_summit.html">signal_at_summit</a></code> </p>
          </td>
          <td><p>Signal around all bedfile midpoints</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_average.html">signal_average</a></code> </p>
          </td>
          <td><p>Calculate average signal by position</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_from_telomeres_average.html">signal_from_telomeres_average</a></code> </p>
          </td>
          <td><p>Compute mean signal and SD by position of output of <code><a href='reference/signal_from_telomeres.html'>signal_from_telomeres</a></code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_from_telomeres.html">signal_from_telomeres</a></code> </p>
          </td>
          <td><p>Collect signal from telomeres for all chromosome arms</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="signal_per_orf_length.html">signal_per_orf_length</a></code> </p>
          </td>
          <td><p>Mean signal for all ORFs genome-wide</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_compress.html">wiggle_compress</a></code> </p>
          </td>
          <td><p>Compress wiggle data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_correlation.html">wiggle_correlation</a></code> </p>
          </td>
          <td><p>Correlation between two ChIP-seq experiments</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_plot.html">wiggle_plot</a></code> </p>
          </td>
          <td><p>Wiggle data line plot</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_scatter_plot.html">wiggle_scatter_plot</a></code> </p>
          </td>
          <td><p>To make a scatter plot comparison of two wiggle files</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_scatter.html">wiggle_scatter</a></code> </p>
          </td>
          <td><p>To make a scatter plot comparison of two wiggle files</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_signal_average.html">wiggle_signal_average</a></code> </p>
          </td>
          <td><p>Calculate genome-wide signal average in wiggle data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="wiggle_smooth.html">wiggle_smooth</a></code> </p>
          </td>
          <td><p>Sliding window smooth of wiggle data</p></td>
        </tr>
      </tbody>
      </table>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale-Silva, Tovah Markowitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
