<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Standard analysis of ChIP-seq experiments — opening_act • hwglabr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Standard analysis of ChIP-seq experiments</h1>
    </div>

    
    <p>This function will run the lab's standard analysis of ChIP-seq experiments, for which
tab-separated wiggle data is generated. It will call different functions in the package
to produce several .pdf files of analysis plots, written to a new folder in
".../LabShare/HTGenomics/Opening_act/".

    <strong>Note:</strong> When running on data aligned to the SK1 genome three of the plots are
not produced: signal at sub-telomeric regions, signal at DSB hotspots and signal at
axis binding sites. The reason for skipping the sub-telomeric signal analysis is the
fact that the SK1 genome annotation contains inconsistencies at sub-telomeric regions.
The reason for skipping the other two analysis is the fact that the reference data for
DSB hotspots and Red1 binding sites were not available for SK1 at the time of writing
this function. It is probably redundant to run this (long!) analysis for both genomes
anyway, and using data mapped to the S288c reference genome should be preferred.</p>
    

    <pre class="usage"><span class='fu'>opening_act</span>(<span class='no'>wiggleData</span>, <span class='no'>relevantGenotype</span>, <span class='no'>chipTarget</span>, <span class='no'>sampleID</span>,
  <span class='kw'>userInput</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>runMetaORF</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>wiggleData</th>
      <td><p>As a list of the 16 chr wiggle data (output of <code><a href='readall_tab.html'>readall_tab</a></code>).
No default.</p></td>
    </tr>
    <tr>
      <th>relevantGenotype</th>
      <td><p>String indicating the relevant strain mutations. Just use "WT",
for example, if there are no relevant mutations. No default.</p></td>
    </tr>
    <tr>
      <th>chipTarget</th>
      <td><p>String indicating the ChIP target protein. No default.</p></td>
    </tr>
    <tr>
      <th>sampleID</th>
      <td><p>String indicating the sample ID, including the ID used in the
analysis pipeline (with a date) and the read mapping conditions (see examples below).
The function asks the user to check that the provided "sampleID" matches the required
format before proceeding with the analysis. No default.</p></td>
    </tr>
    <tr>
      <th>userInput</th>
      <td><p>Boolean indicating whether to ask user to check the format of the
<code>sampleID</code> argument. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>runMetaORF</th>
      <td><p>Boolean indicating whether to run the meta ORF analysis. This analysis
typically takes about 30 minutes to run, so it may be useful to exclude it.
Defaults to <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A new folder in ".../LabShare/HTGenomics/Opening_act/" containing output plots
(as .pdf files) of the following analysis:</p><ol>
<li><p><strong>Chromosome size bias</strong></p></li>
<li><p><strong>Signal at centromeres</strong></p></li>
<li><p><strong>Signal flanking rDNA</strong></p></li>
<li><p><strong>Signal at sub-telomeric regions</strong> (data mapped to S288c reference genome only)</p></li>
<li><p><strong>Signal at DSB hotspots</strong> (data mapped to S288c reference genome only)</p></li>
<li><p><strong>Signal at axis binding sites</strong> (data mapped to S288c reference genome only)</p></li>
<li><p><strong>Signal at meta ORF</strong></p></li>
</ol>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>opening_act</span>(<span class='kw'>wiggleData</span><span class='kw'>=</span><span class='no'>WT</span>, <span class='kw'>relevantGenotype</span><span class='kw'>=</span><span class='st'>"WT"</span>, <span class='kw'>chipTarget</span><span class='kw'>=</span><span class='st'>"Red1"</span>,
            <span class='kw'>sampleID</span><span class='kw'>=</span><span class='st'>"AH119C-040114-sacCer3-2mis"</span>)
<span class='fu'>opening_act</span>(<span class='no'>set1_wiggle_data</span>, <span class='st'>"set1"</span>, <span class='st'>"Red1"</span>, <span class='st'>"AH8584b-16032016-sacCer3-2mis"</span>)
<span class='fu'>opening_act</span>(<span class='no'>rec8</span>, <span class='st'>"rec8"</span>, <span class='st'>"Red1"</span>, <span class='st'>"AH8115b-24042015-SacCer3-2mis"</span>,
            <span class='kw'>userInput</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>runMetaORF</span><span class='kw'>=</span><span class='fl'>FALSE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale-Silva, Tovah Markowitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
