<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper function to run edgeR analysis of summarized reads from RNA-seq experiment — rna_seq_analysis • hwglabr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper function to run edgeR analysis of summarized reads from RNA-seq experiment</h1>
    </div>

    
    <p>This function allows you to run a statistical analysis of an RNA-seq experiment
using Bioconductor package
<a href='https://bioconductor.org/packages/release/bioc/html/edgeR.html'>edgeR</a>.
Inputs are tables of summarized reads generated using
<a href='http://bioinf.wehi.edu.au/featureCounts/'>featureCounts</a> (part of the lab's
RNA-seq analysis pipeline running on NYU's HPC) for each sample in the experiment.
The presence or absence of biological replicates is automatically inferred from the
<code>conditionNames</code> argument: conditions aggregate samples as replicates. If
you provide as many different <code>conditionNames</code> as there are <code>sampleNames</code>
each sample is taken as a single condition (no replicates).
The output includes tables of CPM, TPM and, if included, differential expression
(DE) analysis for selected pairs of samples (see "Value" section below for more
details).

    <strong>Running without replicates:</strong> This function allows you to run the analysis on
experiments without biological replicate libraries. While the calculation of CPM and
TPM has no requirement for replicates, you should obviously avoid doing DE analysis
in the absence of replicates. The statistical model used by
<a href='https://bioconductor.org/packages/release/bioc/html/edgeR.html'>edgeR</a>
to perform DE analysis relies on biological replicates to estimate biological
variability. While there is no satisfactory alternative to actual biological replication,
you can still run a preliminary DE analysis without replicates. The approach
taken here relies on providing an estimate of biological variation that is reasonable
for RNA-seq experiments using <em>S. cerevisiae</em>. This is a better alternative
to simply assuming that biological variability is absent. Typical values for
the common BCV (square root-dispersion) for datasets arising from well-controlled
experiments are 0.1 for data on genetically identical model organisms, so that
value is used here.</p>
    

    <pre class="usage"><span class='fu'>rna_seq_analysis</span>(<span class='no'>pathToFiles</span>, <span class='no'>sampleNames</span>, <span class='no'>conditionNames</span>, <span class='kw'>batchNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>pairwiseDE</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>outputFilePrefix</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pathToFiles</th>
      <td><p>A list of strings corresponding to the full path to the featureCounts
output files for all samples in the experiment. No default.</p></td>
    </tr>
    <tr>
      <th>sampleNames</th>
      <td><p>A list of strings corresponding to the names of all samples
in the experiment in the same order as the files in <code>pathToFiles</code>. No default.</p></td>
    </tr>
    <tr>
      <th>conditionNames</th>
      <td><p>A list of strings corresponding to the experimental groups/conditions
for each sample/library. Will be the input to edgeR's <code><a href='http://www.rdocumentation.org/packages/edgeR/topics/DGEList'>DGEList</a></code>
constructor function and will define replicate groups, if any. No default.</p></td>
    </tr>
    <tr>
      <th>batchNames</th>
      <td><p>Optional list of strings corresponding to the experimental batch for each
sample/library. Will be part of the design matrix for differential expression testing and
will define batch groups, if any. Defaults to <code>NULL</code> (all samples come from the same batch).</p></td>
    </tr>
    <tr>
      <th>pairwiseDE</th>
      <td><p>Logical indicating whether to test differential expression (DE).
Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>outputFilePrefix</th>
      <td><p>Optional string to be added as prefix to output file names.
No default.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The output includes several tables saved as .csv files in a directory
named "RNA-seq_analysis" written to the working directory:</p><ol>
<li><p><strong>CPM:</strong> Compositional bias-normalized <strong>C</strong>ounts <strong>P</strong>er
  <strong>M</strong>illion (output of edgeR's <code><a href='http://www.rdocumentation.org/packages/edgeR/topics/cpm'>cpm</a></code> function). Useful
  for intersample feature expression comparison (not feature length-normalized).</p></li>
<li><p><strong>TPM:</strong> Compositional bias and feature length-normalized <strong>T</strong>ranscripts
  <strong>P</strong>er <strong>M</strong>illion. Useful for within-sample feature expression comparison.</p></li>
<li><p><strong>DE:</strong> A <strong>D</strong>ifferential <strong>E</strong>xpression table comparing the first sample
  to all others. Includes log2 fold change (<code>logFC</code>), average log2 counts per million
  (<code>logCPM</code>), two-sided p-value (<code>PValue</code>) and false discovery rate (<code>FDR</code>).
  <strong>Note:</strong> Fold change differences (<code>logFC</code>) between samples not directly ompared
  in the table can be obtained by subtracting their reported <code>logFC</code> (to the first sample).
  For example, if the first sample is <code>sample1</code> and we want the <code>logFC</code> between
  <code>sample2</code> and <code>sample3</code>, simply calculate the difference:
  <code>logFC.sample3_vs_sample2</code> = <code>logFC.sample3</code> - <code>logFC.sample2</code></p></li>
</ol><p>A <strong>M</strong>ulti-<strong>D</strong>imensional <strong>S</strong>caling plot of all samples is also
saved in the output directory as a .pdf file.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
rna_seq_analysis(pathToFiles = list('AH119-2h_featureCounts.txt', 'AH119-3h_featureCounts.txt',
                                    'AH8104-2h_featureCounts.txt', 'AH8104-3h_featureCounts.txt'),
                 sampleNames = list('AH119_2h', 'AH119_3h', 'AH8104_2h', 'AH8104_3h'),
                 conditionNames = list('WT_2h', 'WT_3h', 'dot1_2h', 'dot1_3h'),
                 outputFilePrefix = 'dot1_noReplicates')

rna_seq_analysis(pathToFiles = list('AH119-2h_featureCounts.txt', 'AH119-3h_featureCounts.txt',
                                    'AH8104-2h_featureCounts.txt', 'AH8104-3h_featureCounts.txt'),
                 sampleNames = list('AH119_2h', 'AH119_3h', 'AH8104_2h', 'AH8104_3h'),
                 conditionNames = list('WT_2h', 'WT_3h', 'dot1_2h', 'dot1_3h'),
                 pairwiseDE = TRUE, outputFilePrefix = 'dot1_noReplicates')

rna_seq_analysis(pathToFiles = list('AH119-A_featureCounts.txt', 'AH119-B_featureCounts.txt',
                                    'AH8104-A_featureCounts.txt', 'AH8104-B_featureCounts.txt'),
                 sampleNames = list('AH119_A', 'AH119_B', 'AH8104_A', 'AH8104_B'),
                 conditionNames = list('WT', 'WT', 'dot1', 'dot1'),
                 batchNames = list('batch1', 'batch2', 'batch1', 'batch2')
                 pairwiseDE = TRUE)
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale-Silva, Tovah Markowitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
