<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Mean signal for all ORFs genome-wide — signal_per_orf_length • hwglabr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Mean signal for all ORFs genome-wide</h1>
    </div>

    
    <p>This function allows you to calculate mean ChIP signal for each ORF in the genome. The function
goes through all included features in the supplied gff data and for each one it adds ORF length
(in bp) and the mean of the signal collected from the supplied ChIP-seq data. This can then be
used to analyse mean signal as a function of ORF length.
The function takes as input the wiggle data as a list of 16 chromosomes (output of
<code><a href='readall_tab.html'>readall_tab</a></code>).

    <strong>Note:</strong> Our wiggle data always contains gaps with missing chromosome coordinates
and ChIP-seq signal. The way this function deals with that is by skipping affected genes.
The number of skipped genes in each chromosome is printed to the console, as well as the
final count (and percentage) of skipped genes.</p>
    

    <pre class="usage"><span class='fu'>signal_per_orf_length</span>(<span class='no'>inputData</span>, <span class='no'>gff</span>, <span class='no'>gffFile</span>, <span class='kw'>saveFile</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inputData</th>
      <td><p>As a list of the 16 chr wiggle data (output of <code><a href='readall_tab.html'>readall_tab</a></code>). No default.</p></td>
    </tr>
    <tr>
      <th>gff</th>
      <td><p>Optional dataframe of the gff providing the ORF cordinates. Must be provided if
<code>gffFile</code> is not. No default. Note: You can use the function <code><a href='gff_read.html'>gff_read</a></code> in hwglabr to
load your selected gff file.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p>Optional string indicating path to the gff file providing the ORF cordinates. Must be
provided if <code>gff</code> is not. No default.</p></td>
    </tr>
    <tr>
      <th>saveFile</th>
      <td><p>Boolean indicating whether output should be written to a .txt file (in current working
directory). If <code>saveFile = FALSE</code>, output is returned to screen or an R object (if assigned).
Defaults to FALSE.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame equivalent to the supplied gff table with the following additional columns:</p><ol>
<li><p><code>length</code> Length of the ORF in bp</p></li>
<li><p><code>mean_signal</code> Mean of the signal collected from the supplied data</p></li>
</ol><p><strong>Note:</strong> Skipped genes are included in the output with 'NA' for <code>mean_signal</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>signal_per_orf_length</span>(<span class='no'>WT</span>, <span class='kw'>gff</span> <span class='kw'>=</span> <span class='no'>gff</span>)

<span class='fu'>signal_per_orf_length</span>(<span class='no'>WT</span>, <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='no'>S288C_annotation_modified.gff</span>, <span class='kw'>saveFile</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale-Silva, Tovah Markowitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
