<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Signal at all ORFs genome-wide (meta ORF) — signal_at_orf • hwglabr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Signal at all ORFs genome-wide (meta ORF)</h1>
    </div>

    
    <p>This function allows you to pull out the ChIP signal over all ORFs in the genome. It collects the
signal over each ORF plus both flanking regions (1/2 the length of the ORF on each side) and
scales them all to the same value (1000). This means that for two example genes with lengths of
500 bp and 2 kb, flanking regions of 250 bp and 1 kb, respectively, will be collected up and
downstream. The whole region is then rescaled to a length of 1000, corresponding to a gene length
of 500 plus 250 for each flanking region. After scaling, a loess model of the signal is built and
used to output predictions of the signal at each position between 1 and 1000. 
The function takes as input the wiggle data as a list of 16 chromosomes.
(output of <code><a href='readall_tab.html'>readall_tab</a></code>).

    <strong>Note:</strong> Our wiggle data always contains gaps with missing chromosome coordinates
and ChIP-seq signal. The way this function deals with that is by skipping affected genes.
The number of skipped genes in each chromosome is printed to the console, as well as the
final count (and percentage) of skipped genes.</p>
    

    <pre class="usage"><span class='fu'>signal_at_orf</span>(<span class='no'>inputData</span>, <span class='no'>gff</span>, <span class='no'>gffFile</span>, <span class='kw'>loessSpan</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>saveFile</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>inputData</th>
      <td><p>As a list of the 16 chr wiggle data (output of <code><a href='readall_tab.html'>readall_tab</a></code>).
No default.</p></td>
    </tr>
    <tr>
      <th>gff</th>
      <td><p>Optional dataframe of the gff providing the ORF cordinates. Must be provided if
<code>gffFile</code> is not. No default. Note: You can use the function <code><a href='gff_read.html'>gff_read</a></code> in
hwglabr to load your selected gff file.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p>Optional string indicating path to the gff file providing the ORF cordinates.
Must be provided if <code>gff</code> is not. No default.</p></td>
    </tr>
    <tr>
      <th>loessSpan</th>
      <td><p>Number specifying <code>span</code> argument for <code>loess</code> function (the smoothing
parameter alpha). This controls the degree of smoothing of the signal.
Defaults to <code>0.05</code>.</p></td>
    </tr>
    <tr>
      <th>saveFile</th>
      <td><p>Boolean indicating whether output should be written to a .txt file (in current
working directory). If <code>saveFile = FALSE</code>, output is returned to screen or an
R object (if assigned). Defaults to <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A local data frame with four columns:</p><ol>
<li><p><code>chr</code> Chromosome number</p></li>
<li><p><code>position</code> Nucleotide coordinate (in normalized total length of 1 kb)</p></li>
<li><p><code>signal</code> ChIP-seq signal at each position (1 to 1000)</p></li>
<li><p><code>gene</code> Systematic gene name</p></li>
</ol>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>signal_at_orf</span>(<span class='no'>WT</span>, <span class='kw'>gff</span> <span class='kw'>=</span> <span class='no'>gff</span>)

<span class='fu'>signal_at_orf</span>(<span class='no'>WT</span>, <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='no'>S288C_annotation_modified.gff</span>,
              <span class='kw'>loessSpan</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>saveFile</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale-Silva, Tovah Markowitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
